### PR TITLE
Fix lp:1661488, Allow to enable userstat=1 for existing connections, keep concurrent connections tracking after FLUSH USER/CLIENT_STATISTICS

### PR DESCRIPTION
**Pull request by ihanick**
Bug fix for https://bugs.launchpad.net/percona-server/+bug/1661488

http://jenkins.percona.com/job/percona-server-5.5-param/1526

For already existing connections tracking required, I've added diff_connections variable for that purpose.
At the same time connections are tracked in two places (USER_STATISTICS and CLIENT_STATISTICS), thus this variable could be bigger than one for each connection.

In addition to enabling userstats in runtime, there is a FLUSH * issue. I've modified corresponding refresh code path in almost the same way as it was in original google patch.

### CODE REVIEW

File: sql/mysqld.h [hunk 1]
```diff
@@ -511,10 +511,13 @@ void init_global_index_stats(void);
 void init_global_client_stats(void);
 void init_global_thread_stats(void);
 void free_global_user_stats(void);
+void free_global_user_stats(void);
```
Comment (laurynas-biveinis) [inline sql/mysqld.h]: Duplicated line

File: sql/sql_class.h [hunk 2]
```diff
@@ -2347,6 +2347,7 @@ class THD :public Statement,
   ulonglong diff_commit_trans, diff_rollback_trans;
   // Number of connection errors not reflected in global_user_stats yet.
   ulonglong diff_denied_connections, diff_lost_connections;
+  ulonglong diff_connections;
```
Comment (laurynas-biveinis) [inline sql/sql_class.h]: Please add a comment

File: sql/sql_connect.cc [hunk 3]
```diff
@@ -184,14 +184,17 @@ void init_user_stats(USER_STATS *user_stats,
...
   DBUG_ENTER("init_user_stats");
-  DBUG_PRINT("info",
-             ("Add user_stats entry for user %s - priv_user %s",
-              user, priv_user));
-  strncpy(user_stats->user, user, sizeof(user_stats->user));
-  strncpy(user_stats->priv_user, priv_user, sizeof(user_stats->priv_user));
-
-  user_stats->user_len=               strlen(user_stats->user);
-  user_stats->priv_user_len=          strlen(user_stats->priv_user);
+  if (user)
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: Why is this check required?

File: sql/sql_connect.cc [hunk 4]
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
...
+
+  if (my_init_dynamic_array(&delete_items, sizeof(USER_STATS *),
+        stats->records, 1))
+  {
+    sql_print_error("reset global stats: cannot initialize dynamic array");
+    my_message(ER_OUTOFMEMORY, ER(ER_OUTOFMEMORY), MYF(0));
+    return;
+  }
+
+  for (uint idx=0;idx < stats->records; idx++)
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: s/uint/ulong, space after ;

File: sql/sql_connect.cc [hunk 5]
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
...
+  {
+    USER_STATS* user_stats=
+      (USER_STATS *) my_hash_element(stats, idx);
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: C++ style cast

File: sql/sql_connect.cc [hunk 6]
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
...
+    if (user_stats->concurrent_connections)
+    {
+      init_user_stats(user_stats, NULL, NULL,
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: This function has way too many zero arguments - this is not the part of this bugfix, but I have to ask you to clean it up anyway: please only keep the arguments that passed here or at the other call in increment_count_by_name, and move the rest of zero initialisation to the function body. Please also make it "static" and remove its prototype from sql/structs.h

File: sql/sql_connect.cc [hunk 7]
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
...
+          0, 0, 0,   // select, update and other commands
+          0, 0,      // commit and rollback trans
+          0,         // diff_denied_connections
+          0,         // lost connections
+          0,         // access denied errors
+          0);        // empty queries
+    }
+    else
+    {
+      if (insert_dynamic(&delete_items, (uchar *) &user_stats))
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: if (unlikely(insert_dynamic ..., C++ style cast

File: sql/sql_connect.cc [hunk 8]
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
   }
 }
 
+void reset_global_user_or_client_stats(HASH* stats)
+{
+  DYNAMIC_ARRAY delete_items;
+
+  if (my_init_dynamic_array(&delete_items, sizeof(USER_STATS *),
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: if (unlikely(my_init ...

File: sql/sql_connect.cc [hunk 9]
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
...
+      {
+        delete_dynamic(&delete_items);
+        sql_print_error("reset global stats: insert_dynamic failed");
+        my_message(ER_OUTOFMEMORY, ER(ER_OUTOFMEMORY), MYF(0));
+        return;
+      }
+    }
+  }
+
+  for (uint i= 0; i < delete_items.elements; ++i)
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: s/uint/ulong

File: sql/sql_connect.cc [hunk 10]
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
...
+  {
+    USER_STATS *entry= *dynamic_element(&delete_items, i, USER_STATS **);
+    (void) my_hash_delete(stats, (uchar*) entry);
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: It's a bug if this my_hash_delete fails - please DBUG_ASSERT its return value; C++-style cast

File: sql/sql_connect.cc [hunk 11]
```diff
@@ -346,6 +349,68 @@ void init_global_index_stats(void)
   }
 }
 
+void reset_global_user_or_client_stats(HASH* stats)
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: static

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 12]
```diff
@@ -71,6 +71,63 @@ connection default;
 --let $assert_cond= "$query_result" = 0
 --source include/assert.inc
 
+# Make a single connection for mysqltest_1 user
+connect (conn1,localhost,mysqltest_1,,);
+SELECT 1;
+connection default;
+
+# Verify that flush statistics keeps concurent connections counter inact
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: intact

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 13]
```diff
@@ -71,6 +71,63 @@ connection default;
 --let $assert_cond= "$query_result" = 0
 --source include/assert.inc
 
+# Make a single connection for mysqltest_1 user
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: The testcase does not actually query the value?

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 14]
```diff
@@ -71,6 +71,63 @@ connection default;
...
+connection default;
+
+# Verify that flush statistics keeps concurent connections counter inact
+FLUSH USER_STATISTICS;
+
+disconnect conn1;
+let $count_sessions= 1;
+--source include/wait_until_count_sessions.inc
+
+# verify flush user statistics with existing connections
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: add connection default here

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 15]
```diff
@@ -71,6 +71,63 @@ connection default;
...
+SET GLOBAL userstat=0;
+# Make a single connection for mysqltest_1 user
+connect (conn1,localhost,mysqltest_1,,);
+SELECT 1;
+connection default;
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: This one is redundant

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by ihanick** [comment]
http://jenkins.percona.com/job/percona-server-5.5-param/1529/
- check connection_concurrency for fault login
- check for count(1) ... where concurrency=0 instead of let $assert_cond= "$query_result" = 0
- rewrite init_user_stats, make it static

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 16]
```diff
@@ -44,33 +44,80 @@ connection default;
 --let $assert_cond= "$query_result" = 1
 --source include/assert.inc
 
-
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: Redundant change?

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 17]
```diff
@@ -44,33 +44,80 @@ connection default;
...
 # if there is no connections, concurrent connections counter should be zero
 disconnect conn1;
 let $count_sessions= 2;
 --source include/wait_until_count_sessions.inc
 
 --let $assert_text= I_S.USER_STATISTICS CONCURRENT_CONNECTIONS: no connections - zero concurrency
---let $query_result= query_get_value(SELECT CONCURRENT_CONNECTIONS FROM INFORMATION_SCHEMA.USER_STATISTICS WHERE USER LIKE 'mysqltest_1', CONCURRENT_CONNECTIONS, 1)
---let $assert_cond= "$query_result" = 0
+--let $query_result= query_get_value(SELECT COUNT(1) cnt FROM INFORMATION_SCHEMA.USER_STATISTICS WHERE USER LIKE 'mysqltest_1' AND  CONCURRENT_CONNECTIONS=0, cnt, 1)
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: Why this change is required?

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 14 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: This comment is not addressed yet (but connection default should be added immediately after disconnect conn1 above)

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 18]
```diff
@@ -44,33 +44,80 @@ connection default;
...
+
+# Verify that flush statistics keeps concurent connections counter intact
+FLUSH USER_STATISTICS;
+
+disconnect conn1;
+let $count_sessions= 1;
+--source include/wait_until_count_sessions.inc
+
+# verify flush user statistics with existing connections
+# connection concurrency tracked only if it runs queries while userstat=1
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: No actual userstat query in this test part?

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 19]
```diff
@@ -44,33 +44,80 @@ connection default;
...
+SELECT 1;
+disconnect conn2;
+disconnect conn1;
+connection default;
+
+let $count_sessions= 1;
+--source include/wait_until_count_sessions.inc
+
+
+# verify flush client statistics with existing connections
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: Likewise?

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 20]
```diff
@@ -44,33 +44,80 @@ connection default;
...
 
---let $assert_text= I_S.USER_STATISTICS CONCURRENT_CONNECTIONS: unknown user connection
---let $query_result= query_get_value(SELECT CONCURRENT_CONNECTIONS FROM INFORMATION_SCHEMA.USER_STATISTICS WHERE USER LIKE 'mysqltest_unknownuser', CONCURRENT_CONNECTIONS, 1)
---let $assert_cond= "$query_result" = 0
+--let $assert_text= I_S.USER_STATISTICS CONCURRENT_CONNECTIONS: wrong password connection
+--let $query_result= query_get_value(SELECT COUNT(1) cnt FROM INFORMATION_SCHEMA.USER_STATISTICS WHERE USER LIKE 'mysqltest_2' AND CONCURRENT_CONNECTIONS=0, cnt, 1)
+--let $assert_cond= "$query_result" = 1
 --source include/assert.inc
 
+# We need alive connection to check that flush will not remove concurent_connections counter
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: No actual userstat query in this test part?

File: sql/sql_connect.cc [hunk 21]
```diff
@@ -158,63 +158,29 @@ void free_thread_stats(THREAD_STATS* thread_stats)
...
-                     ulonglong select_commands,
-                     ulonglong update_commands,
-                     ulonglong other_commands,
-                     ulonglong commit_trans,
-                     ulonglong rollback_trans,
-                     ulonglong denied_connections,
-                     ulonglong lost_connections,
-                     ulonglong access_denied_errors,
-                     ulonglong empty_queries)
+static void init_user_stats(USER_STATS *user_stats)
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: I suggested to remove only the always-zero args, not all the args. The current way is also fine, but then this whole function should be replaced by a memset.

If you choose to restore the non-zero args, then use memset, and then set only those args.

File: sql/sql_connect.cc [hunk 22]
```diff
@@ -346,6 +312,60 @@ void init_global_index_stats(void)
...
+  }
+
+  for (ulong idx=0; idx < stats->records; idx++)
+  {
+    USER_STATS* user_stats=
+      reinterpret_cast<USER_STATS*>(my_hash_element(stats, idx));
+    if (user_stats->concurrent_connections)
+    {
+      uint saved_concurrent_connections = user_stats->concurrent_connections;
+      init_user_stats(user_stats);
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: The need for saved_concurrent_connections would disappear with init_user_stats taking a concurrent_connections arg

File: sql/sql_connect.cc [hunk 23]
```diff
@@ -346,6 +312,60 @@ void init_global_index_stats(void)
...
+        my_message(ER_OUTOFMEMORY, ER(ER_OUTOFMEMORY), MYF(0));
+        return;
+      }
+    }
+  }
+
+  for (ulong i= 0; i < delete_items.elements; ++i)
+  {
+    USER_STATS *entry= *dynamic_element(&delete_items, i, USER_STATS **);
+    DBUG_ASSERT(!my_hash_delete(stats, reinterpret_cast<uchar*>(entry)));
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: This is a bug: assert with a side effect, will get compiled away in a release build. It has to be
#ifndef DBUG_OFF
    whatervertype result=
#endif
    my_hash_delete(...);
    DBUG_ASSERT(!result);

**Comment by laurynas-biveinis** [comment]
Jenkins link is 404?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by ihanick** [comment]
http://jenkins.percona.com/job/percona-server-5.5-param/1537/
- added DBUG_ASSERT for my_hash_delete results
- init_user_stats clears zero fields
- avoid ---let $assert_cond= "$query_result" = 0, because there is no alert for zero rows returned
- changed unknown user connection test to wrong password check, unknown users not creating any record in user stats.
- clarify flush stats test part

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 17 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: OK, I see

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 16 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: This comment has not been addressed

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 20 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: This comment has not been addressed

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 14 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: This comment has not been addressed

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 18 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: This comment has not been addressed

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 19 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: This comment has not been addressed

File: sql/sql_connect.cc (hunk 21 above)
Comment (laurynas-biveinis) [reply sql/sql_connect.cc]: This comment has not been addressed

File: sql/sql_connect.cc (hunk 22 above)
Comment (laurynas-biveinis) [reply sql/sql_connect.cc]: See the comment at init_user_stats

File: sql/sql_connect.cc (hunk 23 above)
Comment (laurynas-biveinis) [reply sql/sql_connect.cc]: This comment has not been addressed

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by ihanick** [comment]
http://jenkins.percona.com/job/percona-server-5.5-param/1546/

- added I_S.USER_STATISTICS and I_S.CLIENT_STATISTICS related to userstats=0 <=> userstats=1 dynamic changes
- Found that it's impossible to track concurrent_connections state while userstats is disabled
- Considered to save state in the same place as in Facebook MySQL: connections are already tracked inside user resources limit
- reverted https://bugs.launchpad.net/percona-server/+bug/728082
- for each show user_statistics or client_statistics query updating number of concurrent connections  by calculating SUM(resource limit concurrent connection) for all users

Benefits:
- smaller change set (diff to original 2791a612d407e8710b690ac1d0b739005bfcec3c )
- correct tracking if user stats was enabled or disabled
- correct flush behaviour (removes all records from statistics)
- same test cases could be used
- codepath is the same for thread pool and one-thread-per-connection, I can merge both test cases to one and eliminate need in additional mysqld restart for mysql-test run

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 16 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: Still not addressed

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 14 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: Still not addressed

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 18 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: OK

File: mysql-test/include/percona_userstat_conn_handling.inc (hunk 19 above)
Comment (laurynas-biveinis) [reply mysql-test/include/percona_userstat_conn_handling.inc]: OK

File: sql/sql_connect.cc [hunk 24]
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
   mysql_mutex_unlock(&LOCK_global_user_client_stats);
 }
 
+static void clear_stats_concurrent_connections(HASH* stats) {
+  for (ulong idx=0; idx < stats->records; idx++)
+  {
+    USER_STATS* user_stats=
+      reinterpret_cast<USER_STATS*>(my_hash_element(stats, idx));
+    user_stats->concurrent_connections = 0;
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: No space before "="

File: sql/sql_connect.cc [hunk 25]
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
...
+  }
+}
+
+static void inc_stats_concurrent_conn(HASH* stats, const char* user_string, int cnt) {
+  USER_STATS* user_stats;
+  if ((user_stats = (USER_STATS *) my_hash_search(stats,
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: Move assignment to var declaration (can make the pointer itself const too)
C++ style cast

File: sql/sql_connect.cc [hunk 26]
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
...
+                                                  (uchar *) user_string,
+                                                  strlen(user_string))))
+  {
+    user_stats->concurrent_connections += cnt;
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: No space before +=

File: sql/sql_connect.cc [hunk 27]
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
...
+  }
+}
+
+// Updates number of concurrent connections for user_stats and client_stats
+// based on account resoure limits
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: "resource"

Comment (laurynas-biveinis) [reply sql/sql_connect.cc]: If a header comment is added (strictly speaking it has to be added for all the new functions, but the previous two are trivial enough so that's not a critical issue), then it has to be a Doxygen one

File: sql/sql_connect.cc [hunk 28]
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
...
+void refresh_concurrent_conn_stats() {
+  mysql_mutex_lock(&LOCK_user_conn);
+
+  mysql_mutex_lock(&LOCK_global_user_client_stats);
+  clear_stats_concurrent_connections(&global_user_stats);
+  clear_stats_concurrent_connections(&global_client_stats);
+  mysql_mutex_unlock(&LOCK_global_user_client_stats);
+
+  for (ulong idx=0; idx < hash_user_connections.records; idx++)
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: space after =

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 29]
```diff
@@ -1,5 +1,7 @@
 SET @userstat_old= @@userstat;
+SET @max_user_connections_old= @@max_user_connections;
 SET GLOBAL userstat=ON;
+SET GLOBAL max_user_connections = 10000;
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: Why is this required?

File: sql/sql_connect.cc [hunk 30]
```diff
@@ -655,6 +649,49 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
...
+  {
+    struct user_conn * uc=
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: const

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by ihanick** [comment]
http://jenkins.percona.com/job/percona-server-5.5-param/1552/
- fixed spacing & brackets
- changed function comment to doxygen one
- const for pointers
- using per-user resource limits
- added connection default after each disconnect

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 31]
```diff
@@ -1,11 +1,16 @@
...
+# concurrent connections count tracked only if resource limits enabled globally
+# or just for specific users
 CREATE USER mysqltest_1@localhost;
+GRANT USAGE ON *.* TO mysqltest_1@localhost WITH MAX_USER_CONNECTIONS 100;
 CREATE USER mysqltest_2@localhost;
+GRANT USAGE ON *.* TO mysqltest_2@localhost WITH MAX_USER_CONNECTIONS 100;
 
 # Make a single connection for mysqltest_1 user
 connect (conn1,localhost,mysqltest_1,,);
+connection conn1;
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: Redundant

File: mysql-test/include/percona_userstat_conn_handling.inc [hunk 32]
```diff
@@ -47,32 +53,106 @@ connection default;
...
+
+let $count_sessions= 1;
+--source include/wait_until_count_sessions.inc
+
 
 DROP USER mysqltest_1@localhost;
 DROP USER mysqltest_2@localhost;
 
-SET GLOBAL userstat= @userstat_old;
+SET GLOBAL userstat= @userstat_old;
```
Comment (laurynas-biveinis) [inline mysql-test/include/percona_userstat_conn_handling.inc]: Missing newline

File: sql/sql_connect.cc [hunk 33]
```diff
@@ -655,6 +649,54 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
...
+    user_stats->concurrent_connections= 0;
+  }
+}
+
+static void inc_stats_concurrent_conn(HASH* stats,
+				      const char* user_string, int cnt)
+{
+  USER_STATS* const user_stats=
+    reinterpret_cast<USER_STATS*>(my_hash_search(stats,
+          (uchar *) user_string,
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: C++-style cast

File: sql/sql_connect.cc [hunk 34]
```diff
@@ -655,6 +649,54 @@ void update_global_user_stats(THD* thd, bool create_user, time_t now)
...
+  mysql_mutex_lock(&LOCK_user_conn);
+
+  mysql_mutex_lock(&LOCK_global_user_client_stats);
+  clear_stats_concurrent_connections(&global_user_stats);
+  clear_stats_concurrent_connections(&global_client_stats);
+  mysql_mutex_unlock(&LOCK_global_user_client_stats);
+
+  for (ulong idx= 0; idx < hash_user_connections.records; idx++)
+  {
+    struct user_conn* const uc=
```
Comment (laurynas-biveinis) [inline sql/sql_connect.cc]: const struct user_conn* const uc=

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
Almost LGTM, cosmetic changes only

**Comment by ihanick** [comment]
http://jenkins.percona.com/job/percona-server-5.5-param/1554/
- c++ cast (and use const uchar*)
- missing new line character at file end
- const pointer and item for userstats iteration

**Review by laurynas-biveinis — APPROVED** [review]

**Comment by ihanick** [comment]
http://jenkins.percona.com/job/percona-server-5.5-param/1556/
There is no dependency for thread pool in current implementation,
removed thread pool test case for userstats
transferred inc file to test file as is.

**Review by laurynas-biveinis — APPROVED** [review]
